### PR TITLE
Add foreach processor documentation

### PR DESCRIPTION
**Pull request by vagimeli**
### Description
Closes content gap

### Issues Resolved
#4647 

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by vagimeli** [comment]
@hdhalter We need a dev to take the first pass at drafting this content. I can support them in refining the content and getting it through the documentation process. The ingest processor template is provided with this PR, so it should help the dev get started. Thanks!

File: _ingest-pipelines/processors/foreach.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+{% include copy-curl.html %}
+
+## Configuration parameters
+
+The following table lists the required and optional parameters for the `foreach` processor.
+
+Parameter | Required/Optional | Description |
+|-----------|-----------|-----------|
+<insert the parameters>
+
```
Comment (vagimeli) [inline _ingest-pipelines/processors/foreach.md:42]: @gaobinlong Please confirm the configuration parameters table is correct for this processor.

File: _ingest-pipelines/processors/foreach.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+Follow these steps to use the processor in a pipeline.
+
+### Step 1: Create a pipeline
+
+The following query creates a pipeline, named <name>, that uses the `foreach` processor to <do what?>: 
+
+```json
+<insert pipeline code example>
+```
```
Comment (vagimeli) [inline _ingest-pipelines/processors/foreach.md:89]: @gaobinlong When I test this above example in Dev Tools, I'm getting the following error. Will you provide a valid pipeline example? 
```json
{
  "error": {
    "root_cause": [
      {
        "type": "parse_exception",
        "reason": "[processors] required property is missing",
        "property_name": "processors"
      }
    ],
    "type": "parse_exception",
    "reason": "[processors] required property is missing",
    "property_name": "processors"
  },
  "status": 400
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+`if` | Optional | A conditional expression to determine whether to execute this processor.
+`on_failure` | Optional	| Specifies how to handle failures for this processor. See the documentation on [Handling failures in pipelines]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
+`ignore_failure` | Optional | If `true`, failures for this processor are ignored. See the documentation on [Handling failures in pipelines]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
+`tag` | Optional | An identifier for this processor. Useful for debugging and metrics.
+
+## Using the processor
+
+Follow these steps to use the processor in a pipeline.
+
+### Step 1: Create a pipeline
```
Comment (vagimeli) [inline _ingest-pipelines/processors/foreach.md:47]: @gaobinlong Please confirm this pipeline example is the best fit for an OpenSearch user.

File: _ingest-pipelines/processors/foreach.md (hunk 1 above)
Comment (gaobinlong) [reply _ingest-pipelines/processors/foreach.md:42]: The common parameter `description` is missing, and could we make the description of all common parameters same among all processors? See https://opensearch.org/docs/latest/ingest-pipelines/processors/rename/.

File: _ingest-pipelines/processors/foreach.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+The following query creates a pipeline named `test-foreach` that uses the `foreach` processor to extract information from a nested JSON structure: 
+
+```json
+PUT _ingest/pipeline/test-foreach
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:52]: Using json processor wrapped in a foreach processor is not a valid case, here's an example which uses the foreach processor to lowercase all the elements in an array.

```
PUT _ingest/pipeline/test-foreach
{
  "description": "Lowercase all the elements in an array",
  "processors": [
    {
      "foreach": {
        "field": "protocols",
        "processor": {
          "lowercase": {
            "field": "_ingest._value"
          }
        }
      }
    }
  ]
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+### Step 2 (Optional): Test the pipeline
+
+It is recommended that you test your pipeline before you ingest documents.
+{: .tip}
+
+To test the pipeline, run the following query:
+
+```json
+<insert code example>
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:80]:
```
POST _ingest/pipeline/test-foreach/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "protocols": ["HTTP","HTTPS","TCP","UDP"]
      }
    }
  ]
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+```
+{% include copy-curl.html %}
+
+#### Response
+
+The following example response confirms that the pipeline is working as expected:
+
+```json
+{
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:89]:
```

{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "protocols": [
            "http",
            "https",
            "tcp",
            "udp"
          ]
        },
        "_ingest": {
          "_value": null,
          "timestamp": "2024-05-23T02:44:10.8201Z"
        }
      }
    }
  ]
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+}
+```
+{% include copy-curl.html %}
+
+### Step 3: Ingest a document 
+
+The following query ingests a document into an index named `testindex1`:
+
+```json
+PUT testindex1/_doc/1?pipeline=test-foreach
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:121]:
```
POST testindex1/_doc/1?pipeline=test-foreach
{
  "protocols": ["HTTP","HTTPS","TCP","UDP"]
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+}
+```
+{% include copy-curl.html %}
+
+#### Response
+
+The request indexes the document into the index `testindex1` and indexes all documents with the extracted JSON data from the `users` field:
+
+```json
+{
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:136]:
```
{
  "_index": "testindex1",
  "_id": "1",
  "_version": 6,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 5,
  "_primary_term": 67
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+GET testindex1/_doc/1
+```
+{% include copy-curl.html %}
+
+#### Response
+
+The response shows the document with the extracted JSON data from the `users` field:
+
+```json
+{
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:166]:
```
{
  "_index": "testindex1",
  "_id": "1",
  "_version": 6,
  "_seq_no": 5,
  "_primary_term": 67,
  "found": true,
  "_source": {
    "protocols": [
      "http",
      "https",
      "tcp",
      "udp"
    ]
  }
}
```

File: _ingest-pipelines/processors/foreach.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
+---
+layout: default
+title: Foreach
+parent: Ingest processors
+nav_order: 110
+---
+
+# Foreach processor
+
+The `foreach` processor is used to iterate over a list of values in an input document and perform some operation on each value. This can be useful for tasks like extracting information from a nested JSON structure or applying transformations to a collection of fields.
```
Comment (gaobinlong) [inline _ingest-pipelines/processors/foreach.md:10]: Extracting information from a nested JSON structure is not a valid case, the most important use case is to process all the elements in an array in the same way, for example, lowercase or uppercase all the elements in a string array.

File: _ingest-pipelines/processors/foreach.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+## Configuration parameters
+
+The following table lists the required and optional parameters for the `foreach` processor.
+
+Parameter | Required/Optional | Description |
+|-----------|-----------|-----------|
+`field` | Required | The array field to iterate over.
+`processor` | Required | The processor to execute against each field.
+`ignore_missing` | Optional | If `true` and the specified field does not exist or is null, the processor will quietly exit without modifying the document.
```
[the remark this answers is not in the document]
Comment (vagimeli) [reply _ingest-pipelines/processors/foreach.md]: It's a technical expression that means hat the processor will terminate or exit its execution without performing any further actions or modifications to the document and without producing any errors or notifications.

**Pull request merged**